### PR TITLE
persist: fix seal operator to only downgrade after successful seals

### PR DESCRIPTION
**Pull request by ruchirK**
### Motivation

This PR adds a correctness test for the `seal` operator similar to the one we have for `conditional_seal`. It also fixes the `seal` operator to properly pass that test.

It also:

- teaches the `seal` operator to retry on failed attempts to `seal`
- gets rid of the error stream on the `seal` operator

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR adds a release note for any [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/user/content/release-notes.md#what-changes-require-a-release-note).

### CODE REVIEW

**Comment by ruchirK** [comment]
This extracts the first commit from #9417 with the test, and fixes the `seal` operator whilst still using `build_async`. The code now doesn't populate an error stream, and retries on error.

I think this version is nice and simple, but I _think_ the version with explicit futures management is going to be better, because we can batch up multiple seals into a single write out to storage.

File: src/persist/src/operators/stream.rs [hunk 1]
```diff
@@ -287,68 +269,44 @@ where
...
+                            &operator_name,
+                            frontier_element,
+                            e
+                        );
 
-                            let mut session = error_output.session(err_cap);
-                            // TODO: Make error retractable? Probably not...
-                            session.give((e.to_string(), *err_cap.time(), 1));
-                        }
+                        return;
```
Comment (ruchirK) [inline src/persist/src/operators/stream.rs:292]: add a comment explaining that if this fails, we will simply retry the next time the operator is invoked?

**Comment by aljoscha** [comment]
This PR is still the version with `async_op!`, are you planning to change that back to the "by hand" version before we merge?

But yes, I agree that the manual version is potentially better, even though it's somewhat more complicated.

**Comment by aljoscha** [comment]
Gentle ping. :point_up:

**Comment by ruchirK** [comment]
I'm really sorry aljoscha! I dunno why but I'm seeing this now for the first time. Yeah I was planning on merging this, and then doing the "change `seal` to manually track futures" and "change conditional seal to be implemented in terms of `seal`" in separate commits in the other pr.

how does that sound?

File: src/persist/src/operators/stream.rs [hunk 2]
```diff
@@ -288,68 +270,46 @@ where
...
-                        if let Err(e) = result {
-                            log::error!(
-                                "Error sealing {} up to {}: {:?}",
-                                &operator_name,
-                                frontier_element,
-                                e
-                            );
+                    if let Err(e) = write.seal(*frontier_element).await {
+                        // If we fail to seal, simply exit and try again the next
+                        // time the operator is scheduled.
```
Comment (aljoscha) [inline src/persist/src/operators/stream.rs:287]: Do we control when the operator is scheduled again?

**Review by aljoscha — APPROVED** [review]
I had one inline question? The code seems good, though. The test is just a copy of the conditional version, right?

**Comment by aljoscha** [comment]
> I'm really sorry aljoscha! I dunno why but I'm seeing this now for the first time.

No worries at all! :hugs: 

> Yeah I was planning on merging this, and then doing the "change seal to manually track futures" and "change conditional seal to be implemented in terms of seal" in separate commits in the other pr.

Sounds good!

File: src/persist/src/operators/stream.rs (hunk 2 above)
Comment (ruchirK) [reply src/persist/src/operators/stream.rs:287]: No, we don't. but the operator will be rescheduled if the frontier advances again or if theres new data

Comment (ruchirK) [reply src/persist/src/operators/stream.rs:293]: pushed a new commit lmk what you think

Comment (aljoscha) [reply src/persist/src/operators/stream.rs:293]: That seems good, yes.

**Pull request merged**